core/include/scipp/core/sizes.h

```diff
@@ -8,49 +8,80 @@
 
 #include "scipp-core_export.h"
 #include "scipp/common/index.h"
-#include "scipp/core/dimensions.h"
+#include "scipp/common/span.h"
 #include "scipp/core/slice.h"
 #include "scipp/units/dim.h"
 
 namespace scipp::core {
 
-/// Sibling of class Dimensions, but unordered.
-class SCIPP_CORE_EXPORT Sizes {
-public:
-  Sizes() = default;
-  Sizes(const Dimensions &dims);
-  Sizes(const std::unordered_map<Dim, scipp::index> &sizes) : m_sizes(sizes) {}
+constexpr int32_t NDIM_MAX = 6;
 
-  bool operator==(const Sizes &other) const;
-  bool operator!=(const Sizes &other) const;
+class Dimensions;
 
-  bool contains(const Dim dim) const noexcept {
-    return m_sizes.count(dim) != 0;
+/// Small (fixed maximum size) and stable (preserving key order) map
+template <class Key, class Value, int16_t Capacity>
+class SCIPP_CORE_EXPORT small_stable_map {
+public:
+  static constexpr auto capacity = Capacity;
+
+  small_stable_map() = default;
+
+  bool operator==(const small_stable_map &other) const noexcept;
+  bool operator!=(const small_stable_map &other) const noexcept;
+
+  auto begin() const noexcept { return m_keys.begin(); }
+  auto end() const noexcept { return m_keys.begin() + size(); }
+  typename std::array<Key, Capacity>::const_iterator find(const Key &key) const;
+  [[nodiscard]] constexpr bool empty() const noexcept { return size() == 0; }
+  constexpr scipp::index size() const noexcept { return m_size; }
+  bool contains(const Key &key) const noexcept;
+  scipp::index index(const Key &key) const;
+  const Value &operator[](const Key &key) const;
+  const Value &at(const Key &key) const;
+  void assign(const Key &key, const Value &value);
+  void insert_left(const Key &key, const Value &value);
+  void insert_right(const Key &key, const Value &value);
+  void erase(const Key &key);
+  void clear() noexcept;
+  void replace_key(const Key &from, const Key &to);
+  constexpr scipp::span<const Key> keys() const &noexcept {
+    return {m_keys.data(), static_cast<size_t>(size())};
   }
-
-  scipp::index count(const Dim dim) const noexcept {
-    return m_sizes.count(dim);
+  constexpr scipp::span<const Value> values() const &noexcept {
+    return {m_values.data(), static_cast<size_t>(size())};
   }
 
-  auto begin() const { return m_sizes.begin(); }
-  auto end() const { return m_sizes.end(); }
+private:
+  int16_t m_size{0};
+  std::array<Key, Capacity> m_keys{};
+  std::array<Value, Capacity> m_values{};
+};
```

Comment:
It can easily be moved to another (cpp) file? All you need is an explicit instantiation.

The merit is enabling us to do so easily if required.

The main hurdle for making it generic is finding a nice way of having custom exception types (easy using a template arg) and texts (not so easy?). Since we don't have an actual second application for this class at this point I felt it was not worth the effort.

Reply:
Why do we go through the effort of having this as a separate templated type? Should we at least drop the template argument given that you can't use this outside of the Sizes context?

Reply:
You *can* use it outside the `Sizes` context with small effort (changing the way exceptions are written). I am not sure why I should undo the effort now, just because we are only using this in one place so far, which would be even more effort? And even more effort if we ever *do* want to use it ouside, to add the template args back in.